### PR TITLE
add unit tests for json

### PR DESCRIPTION
**Pull request by Jim8y**
**This pr was created one year ago, now ready to merge.** 

This pr will focus on adding unit tests for neo.json to increase the coverage.

Why no issue for this?

this pr focus on things that is very straightforward, and our guidline for coredev is for new core protocol.
Adding unit tests for neo-core is part of the roadmap originally planed after the monorepo, but it will take a very long time to have a monorepo, i will directly work on unit tests in parallel.

What will this pr cover?

this pr will only focus on adding unit tests based on the coverage reports.
The tests will only focus on the Neo.json project and will not touch the logic code of the Neo.json.
This pr will not change existing unit tests, but will only add new test cases.

How to review?

new test cases will be automatically executed by the test workflow.

When to review?

I will keep working to add new tests, no need to review while pr is in draft mode. Review after its ready.

Issues discovered:

```C#
            // Test with an empty JSON Path
            // Assert.ThrowsException<FormatException>(() => json.JsonPath(""));
            // Test with invalid array indexing
            // Assert.ThrowsException<FormatException>(() => json.JsonPath("$.store.book['one']"));
            // Assert.ThrowsException<FormatException>(() => json.JsonPath("$.store.book[999]"));
            // Test with incorrect use of wildcards
            // Assert.ThrowsException<FormatException>(() => json.JsonPath("$.*.store"));
            // Test with invalid use of wildcard in array index
            // Assert.ThrowsException<FormatException>(() => json.JsonPath("$.store.book[*]"));
            // Test with misplaced recursive descent operator
            // Assert.ThrowsException<FormatException>(() => json.JsonPath("$..store..book"));
```

### CODE REVIEW

**Comment by vncoelho** [comment]
Very good @Liaojinghui

**Comment by Liaojinghui** [comment]
@shargon grand master i need your help, test case TestComparisonsWithOtherBooleans in UT_JBoolean get error `Assert.AreEqual failed. Expected:<true>. Actual:<true>. `  i am totally lost here, please help me.

**Comment by shargon** [comment]
I miss this notification, sorry, I will check it soon

**Comment by shargon** [comment]
> @shargon grand master i need your help, test case TestComparisonsWithOtherBooleans in UT_JBoolean get error `Assert.AreEqual failed. Expected:<true>. Actual:<true>. ` i am totally lost here, please help me.
> 
> To repeat:
> 
> line 54 => `Assert.AreEqual(jTrue, new JBoolean(true));`

I think that because they are printed, they show the same, but the reference are different, you should use a method that call IEqualityComparer, like `Assert.IsTrue(jTrue.Equals(new JBoolean(true)));`

**Comment by Jim8y** [comment]
Close as inactive

File: tests/Neo.Json.UnitTests/UT_JBoolean.cs [hunk 1]
```diff
@@ -9,6 +9,8 @@
 // Redistribution and use in source and binary forms with or without
 // modifications are permitted.
 
+using Newtonsoft.Json;
```
Comment (shargon) [inline tests/Neo.Json.UnitTests/UT_JBoolean.cs:12]: When is used?

Comment (Jim8y) [reply tests/Neo.Json.UnitTests/UT_JBoolean.cs:12]: TestSerializationAndDeserialization

**Review by cschuchardt88 — APPROVED** [review]

**Review by NGDAdmin — APPROVED** [review]

**Review by shargon — COMMENTED** [review]
LGTM

**Pull request merged**